### ISSUE TITLE
dynamics assignment issue

### ISSUE THREAD

**Issue by kenmcmi** [AdhocAdam/smletsexchangeconnector#151]
Testing the Dynamic Assignment in 1.6

I've set the value to random in the script

The email is processed but does not assign an analyst based off the support. This can be done manually in the Cireson portal.

On 9.4.1 of the portal 

Attached are the errors that are outputted in the powershell console when it processes an email
[error.txt](https://github.com/AdhocAdam/smletsexchangeconnector/files/3674027/error.txt)

**Comment by AdhocAdam** [comment]
After looking at the attached error output...


**Error 1**
This is related to the multi-mailbox feature and appears to be identical to #141 
```powershell
PS C:\scripts\1.6> C:\scripts\1.6\smletsExchangeConnector.ps1
Index operation failed; the array index evaluated to null.
At C:\scripts\1.6\smletsExchangeConnector.ps1:2450 char:24
+                 return $Mailboxes[$ScsmEmail]
+                        ~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : NullArrayIndex
```

**Error 2**
This is related directly to Dynamic Analyst Assignment. A few things here, while Get-SCSMObject is the first "break" since it's missing a required parameter to identify the Support Group from the Template to the Cireson Support Group Mapping. It seems _even if_ it were to work in this context, it would immediately fail since Get-AdDomain isn't identified as a valid command. In which case, this looks like your missing the Active Directory PowerShell module wherever the connector is running. The failure to leverage AD PowerShell then leads to the final issue wherein Get-Random doesn't have a valid input. But again - this all starts with the missing value. Assuming you are using Incidents as the default new work item and not using Azure Machine Learning. Let's breakpoint the following line and see what kind of values are getting produced and how they ultimately pass into Set-AssignedToPerSupportGroup function as this calls the Get-TierMembers function where we're seeing the error you've attached.

https://github.com/AdhocAdam/smletsexchangeconnector/blob/fb8c14198bc24741fedc2c91915ccdec2f0e0b71/smletsExchangeConnector.ps1#L832

```powershell
Get-SCSMObject : Cannot bind argument to parameter 'Id' because it is null.
At C:\scripts\1.6\smletsExchangeConnector.ps1:1841 char:54
+     $grpInScsm = (Get-ScsmObject @scsmMGMTParams -Id $groupId)
+                                                      ~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Get-SCSMObject], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationErrorNullNotAllowed,SMLets.GetSMObjectCommand
 
Get-AdDomain : The term 'Get-AdDomain' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, 
verify that the path is correct and try again.
At C:\scripts\1.6\smletsExchangeConnector.ps1:1845 char:16
+     $AdRoot = (Get-AdDomain @adParams -Identity $grpInScsm.Domain).DNSRoot
+                ~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (Get-AdDomain:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException
 
Get-Random : Cannot validate argument on parameter 'InputObject'. The argument is null or empty. Provide an argument that is not null or empty, and then try the command again.
At C:\scripts\1.6\smletsExchangeConnector.ps1:1886 char:52
+             $userToAssign = $supportGroupMembers | Get-Random
+                                                    ~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Get-Random], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationError,Microsoft.PowerShell.Commands.GetRandomCommand
```

**Error 3**
This could be coming from a couple places such as setting attempting to set Announcements on the Cireson Portal or retrieving KA/RO Suggestions from the Cireson portal as it's attempting to retrieve a user from the Cireson Portal.
```powershell
Invoke-RestMethod : {"success":false,"message":"User is not authenticated.","exception":null}
At C:\scripts\1.6\smletsExchangeConnector.ps1:2093 char:36
+         $ciresonPortalUserObject = Invoke-RestMethod -Uri ($ciresonPortalServer+ ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-RestMethod], WebException
    + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeRestMethodCommand

PS C:\scripts\1.6> 
```

**Comment by AdhocAdam** [comment]
I've been looking these functions up and down, I've tried them across different environments, and they are no less from the current v1.6. I've only made two modifications here -

1. You have to set the Display Name of your Incident Template as you've configured in the connector
2. I've removed splatting from the functions, so you'll need to run this wherever you're currently running the SMlets Exchange Connector.

The output of this should return 2 users from the Support Group as defined in an Incident Template, with an Enum to AD Group mapping as defined in the Cireson Portal Mappings setting. If these don't work, I'd ask you breakpoint in either ISE/Code to see the exact moment a null value is returned.

```powershell
$irTemplate = Get-SCSMObjectTemplate -DisplayName "Default Email Incident Template Display Name Here"
$domainUserClass = Get-SCSMClass -name "System.Domain.User$"
$assignedToUserRelClass  = Get-SCSMRelationshipClass -name "System.WorkItemAssignedToUser$"

function Get-TierMembers ($TierEnumId)
{
    #define classes
    $mapCls = Get-ScsmClass  -Name "Cireson.SupportGroupMapping"

    #pull the group based on support tier mapping
    $mapping = $mapCls | Get-ScsmObject  | ? { $_.SupportGroupId.Guid -eq $TierEnumId }
    $groupId = $mapping.AdGroupId

    #get the AD group object name
    $grpInScsm = (Get-ScsmObject  -Id $groupId)
    $grpSamAccountName = $grpInScsm.UserName
    
    #determine which domain to query, in case of multiple domains and trusts
    $AdRoot = (Get-AdDomain  -Identity $grpInScsm.Domain).DNSRoot

    if ($grpSamAccountName)
    {
        # Get the group membership
        [array]$supportTierMembers = Get-ADGroupMember  -Server $AdRoot -Identity $grpSamAccountName -Recursive | foreach-object {Get-SCSMObject -Class $domainUserClass -filter "Username -eq '$($_.samaccountname)'"}
    }
    return $supportTierMembers
}

function Get-AssignedToWorkItemVolume ($SCSMUser)
{
    #initialize the counter, get the user's assigned Work Items that aren't in some form of "Done"
    $assignedCount = 0
    $assignedWorkItemRelationships = Get-SCSMRelationshipObject -TargetRelationship $assignedToUserRelClass -TargetObject $SCSMUser 
    $assignedWorkItemRelationships = $assignedWorkItemRelationships | select-object SourceObject -ExpandProperty SourceObject | select-object -ExpandProperty values | ?{($_.type.name -eq "Status") -and (($_.value -notlike "*Resolve*") -and ($_.value -notlike "*Close*") -and ($_.value -notlike "*Complete*") -and ($_.value -notlike "*Skip*") -and ($_.value -notlike "*Cancel*"))}
    $assignedWorkItemRelationships | foreach-object {$assignedCount++}
    
    #build Assigned To Volume object
    $assignedToVolume = New-Object System.Object
    $assignedToVolume | Add-Member -type NoteProperty -name SCSMUser -value $SCSMUser
    $assignedToVolume | Add-Member -type NoteProperty -name AssignedCount -value $assignedCount
    return $assignedToVolume
}

#support group members
$tier = $irTemplate | select-object -expandproperty propertycollection | where-object{($_.path -like "*TierQueue*")} | select-object -ExpandProperty mixedvalue
$supportGroupMembers = Get-TierMembers -TierEnumId $tier

#random analyst
$supportGroupMembers | Get-Random
#least assigned analyst
$supportGroupMembers | foreach-object {Get-AssignedToWorkItemVolume -SCSMUser $_} | Sort-Object AssignedCount -Descending | Select-Object SCSMUser -ExpandProperty SCSMUser -first 1
```

**Comment by kenmcmi** [comment]
Here are the errors

Kenneth McMichael
Manager, IT/Helpdesk
[cid:image001.png@01D59EBD.E7A25240]
(p) 800-638-9797 ext:#2875
(m) 812-746-0911
kenmcmi@atlasworldgroup.com
www.atlasvanlines.com
1212 St. George Road • Evansville, IN 47711

OUR MISSION:
Every day we help you go new places,
with expertise in moving and logistics,
bringing the world within reach.

**Comment by kenmcmi** [comment]
Breakpooints
https://drive.google.com/open?id=1catGhspSAjuaUlfFT0yDGZpXnSGAnSWl

Line by line
https://drive.google.com/open?id=1i3HfpMsCM3eC5GAk9SnsrEqOFJRgo3WV


Kenneth McMichael
Manager, IT/Helpdesk
[cid:image001.png@01D59ED4.1ABD8120]
(p) 800-638-9797 ext:#2875
(m) 812-746-0911
kenmcmi@atlasworldgroup.com
www.atlasvanlines.com
1212 St. George Road • Evansville, IN 47711

OUR MISSION:
Every day we help you go new places,
with expertise in moving and logistics,
bringing the world within reach.

**Comment by ckizziar** [comment]
Well @AdhocAdam not sure what the bounty board is but I was able to reproduce this in 2.1 and 2.2 so here goes nothing. It seems depending on which default template is used, expanding the properties either presented no TierQueue

```
PS C:\ManagementPacks\smletsexchangeconnector-2.1> $irTemplate | select-object -expandproperty propertycollection
Path                                                                        MixedValue                                                                  
----                                                                        ----------                                                                  
$Context/Property[Type='WorkItemIncident!System.WorkItem.Incident']/Source$ $MPElement[Name='WorkItemIncident!IncidentSourceEnum.Portal']$              
$Context/Property[Type='WorkItem!System.WorkItem.TroubleTicket']/Impact$    $MPElement[Name='WorkItem!System.WorkItem.TroubleTicket.ImpactEnum.Medium']$
```
or returned one of the OOB default enumerations, in this case 'Tier1' which had been renamed 'T1 - Helpdesk'

```
PS C:\ManagementPacks\smletsexchangeconnector-2.1> $irTemplate | select-object -expandproperty propertycollection
Path                                                                      MixedValue                                                                   
----                                                                      ----------                                                                   
$Target/Property[Type='WorkItem!System.WorkItem.TroubleTicket']/Impact$   $MPElement[Name='WorkItem!System.WorkItem.TroubleTicket.ImpactEnum.Medium']$ 
$Target/Property[Type='WorkItem!System.WorkItem.TroubleTicket']/Urgency$  $MPElement[Name='WorkItem!System.WorkItem.TroubleTicket.UrgencyEnum.Medium']$
$Target/Property[Type='CoreIncident!System.WorkItem.Incident']/TierQueue$ $MPElement[Name='IncidentTierQueuesEnum.Tier1']$

```
Using a custom template as the default IR template, with a default support group set to 'T1 - Helpdesk' resolved the issue, properly returning the group ID for the TierQueue$ property.


@kenmcmi - I was able to reproduce this issue with the SCSM default templates "Default Incident Template" and "Generic Incident Request" as well as custom templates that did not have anything populated in the support group field by default, but the Cireson Support Group Mapping was set up correctly. You can't change it on defaults, but when creating a custom template you can populate this field on the form and save it to maintain a default value for that group. Can you try these steps to see if that resolves the problem on your end as well?

First, either create a custom incident template or edit the one you're using if it's not part of the sealed defaults, on the properties page (if editing) make sure the box is checked to open the template form after clicking ok, and click aforementioned ok.
![image](https://user-images.githubusercontent.com/11651027/83566822-38c3a280-a4e6-11ea-8be8-96c30d23f8c1.png)

Once the form loads, populate the support group field with your desired default group and click ok.
![image](https://user-images.githubusercontent.com/11651027/83567066-a1128400-a4e6-11ea-999c-5b0a77bcfa70.png)

**Comment by AdhocAdam** [comment]
Thank you for taking the time to confirm this functionality @ckizziar!

I hope it's something this simple in which case it's just a matter of updating the documentation to ensure those using the connector are using wholly custom templates.

**Comment by AdhocAdam** [comment]
Wanted to bring up the recently addressed #188 here as well. But it should be noted it was centered around an actual issue with the logic to pick the analyst rather than being wholly an issue with the feature itself. Adding/citing this as further evidence to @ckizziar's point above.

**Comment by AdhocAdam** [comment]
Per documentation posted by @ckizziar and issues/pull requests raised by @bennyguk - this issue is being closed.
